# 📉 Telco Customer Churn Prediction

## 🧠 Business Problem

Customer churn is one of the biggest challenges for telecom companies. Acquiring a new customer is significantly more expensive than retaining an existing one. Therefore, identifying customers who are likely to churn allows companies to take proactive actions and reduce revenue loss.

---

## 🎯 Objective

The objective of this project is to build a machine learning model that can predict whether a customer is likely to churn based on historical customer data.

---

## 💼 Business Impact

This model can help the business:

- Identify high-risk customers before they leave  
- Enable targeted retention campaigns  
- Optimize marketing spend by focusing on the right customers  
- Improve overall customer lifetime value (CLTV)

---

## 📌 Problem Type

- Binary Classification Problem  
- Target Variable: **Churn (Yes / No)**

---

## 📚 Step 1: Import Required Libraries

In this section, we import all the necessary libraries for data manipulation, visualization, and model building.

In [70]:
# Core libraries for data manipulation and visualization

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## 📥 Step 2: Load Dataset

In this step, we load the Telco Customer Churn dataset and perform an initial inspection to understand its structure.

We will examine:
- Number of rows and columns  
- Sample records  
- Basic structure of the dataset  

In [71]:
# Load dataset
df = pd.read_excel("../dataset/Telco_customer_churn.xlsx")

# Preview first few rows
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,No,No,Yes,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,No,No,Yes,8,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,No,Yes,Yes,28,Yes,Yes,Fiber optic,No,No,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,No,No,Yes,49,Yes,Yes,Fiber optic,No,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


In [72]:
# Check dataset shape
df.shape

(7043, 33)

### 📊 Observation

- Dataset contains **7043 rows and 33 columns**
- Each row represents a unique customer

## 🔍 Step 3: Data Understanding

In this section, we analyze the dataset to understand:

- Feature types  
- Unique values  
- Redundant columns  
- Potential data issues  

This helps in deciding what to keep, drop, or transform.

In [73]:
pd.set_option('display.max_columns', None)
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,No,No,Yes,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,No,No,Yes,8,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,No,Yes,Yes,28,Yes,Yes,Fiber optic,No,No,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,No,No,Yes,49,Yes,Yes,Fiber optic,No,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


In [74]:
df.shape

(7043, 33)

In [75]:
df.describe()

,Count,Zip Code,Latitude,Longitude,Tenure Months,Monthly Charges,Churn Value,Churn Score,CLTV
count,7043.0,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,1.0,93521.964646,36.282441,-119.798880,32.371149,64.761692,0.265370,58.699418,4400.295755
std,0.0,1865.794555,2.455723,2.157889,24.559481,30.090047,0.441561,21.525131,1183.057152
min,1.0,90001.000000,32.555828,-124.301372,0.000000,18.250000,0.000000,5.000000,2003.000000
25%,1.0,92102.000000,34.030915,-121.815412,9.000000,35.500000,0.000000,40.000000,3469.000000
50%,1.0,93552.000000,36.391777,-119.730885,29.000000,70.350000,0.000000,61.000000,4527.000000
75%,1.0,95351.000000,38.224869,-118.043237,55.000000,89.850000,1.000000,75.000000,5380.500000
max,1.0,96161.000000,41.962127,-114.192901,72.000000,118.750000,1.000000,100.000000,6500.000000


In [76]:
df.isnull().sum()

CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

In [77]:
# Identify columns with only one unique value
for col in df.columns:
    if df[col].nunique() == 1:
        print(f"{col}: {df[col].unique()}")

Count: [1]
Country: ['United States']
State: ['California']


### 📊 Observation 1: Constant Columns

The following columns have only one unique value across all rows:

- Count  
- Country  
- State  

These columns do not provide any useful information for modeling and can be removed.

In [78]:
df['Zip Code'].nunique()

1652

In [79]:
df.groupby('City')['Zip Code'].nunique()

City
Acampo          1
Acton           1
Adelanto        1
Adin            1
Agoura Hills    1
               ..
Yreka           1
Yuba City       2
Yucaipa         1
Yucca Valley    1
Zenia           1
Name: Zip Code, Length: 1129, dtype: int64

### 📊 Observation 2: Location Features

- Most cities are associated with only one zip code  
- Some large cities (e.g., Los Angeles, San Diego) have multiple zip codes  

This indicates that:
- Zip Code may contain more granular information than City  
- However, both may introduce high cardinality  

These features should be handled carefully or excluded during modeling.

In [80]:
df.dtypes

CustomerID            object
Count                  int64
Country               object
State                 object
City                  object
Zip Code               int64
Lat Long              object
Latitude             float64
Longitude            float64
Gender                object
Senior Citizen        object
Partner               object
Dependents            object
Tenure Months          int64
Phone Service         object
Multiple Lines        object
Internet Service      object
Online Security       object
Online Backup         object
Device Protection     object
Tech Support          object
Streaming TV          object
Streaming Movies      object
Contract              object
Paperless Billing     object
Payment Method        object
Monthly Charges      float64
Total Charges         object
Churn Label           object
Churn Value            int64
Churn Score            int64
CLTV                   int64
Churn Reason          object
dtype: object

### 📊 Observation 3: Data Types

- Dataset contains both categorical and numerical features  
- Some numerical-looking columns may require type conversion  
- Target variable: **Churn Label / Churn Value**

### 🚨 Key Findings from Data Understanding

- Some columns are irrelevant (CustomerID, Count, Country, State)  
- Location-based features have high cardinality  
- Dataset contains both categorical and numerical features  
- Target variable is available in both label and numeric format  

Next step: Data Cleaning and Preprocessing

---

## 🧹 Step 4: Data Cleaning & Feature Selection

In this step, we clean the dataset by:

- Removing irrelevant and redundant features  
- Handling data type inconsistencies  
- Preparing the dataset for analysis and modeling  

The goal is to retain only meaningful features that contribute to predicting churn.

### 📊 Feature Removal Justification

The following columns were removed:

- **CustomerID** → Unique identifier, no predictive value  
- **Count** → Constant column (no variation)  
- **Country, State** → Single value across dataset  
- **Lat Long** → Redundant (Latitude & Longitude already exist)

Removing these helps reduce noise and improves model efficiency.

In [81]:
# Drop irrelevant and redundant columns
df = df.drop(columns=[
    'CustomerID',
    'Count',
    'Country',
    'State',
    'Lat Long'
])

### 🎯 Target Variable Selection

- **Churn Value (0 / 1)** is used as the target variable  
- Other churn-related columns are removed to avoid data leakage  

Final Target: **Churn**

In [82]:
# Use numeric target variable
df['Churn'] = df['Churn Value']

# Drop redundant churn columns
df = df.drop(columns=['Churn Label', 'Churn Value', 'Churn Score', 'Churn Reason'])

### 📊 Data Type Correction

- `Total Charges` was stored as an object type  
- Converted to numeric format for proper analysis  

Missing values introduced during conversion will be handled next.

In [83]:
# Convert Total Charges to numeric
df['Total Charges'] = pd.to_numeric(df['Total Charges'], errors='coerce')

In [84]:
# Check missing values
df.isnull().sum()

City                  0
Zip Code              0
Latitude              0
Longitude             0
Gender                0
Senior Citizen        0
Partner               0
Dependents            0
Tenure Months         0
Phone Service         0
Multiple Lines        0
Internet Service      0
Online Security       0
Online Backup         0
Device Protection     0
Tech Support          0
Streaming TV          0
Streaming Movies      0
Contract              0
Paperless Billing     0
Payment Method        0
Monthly Charges       0
Total Charges        11
CLTV                  0
Churn                 0
dtype: int64

In [85]:
# Fill missing values in Total Charges with median
df['Total Charges'].fillna(df['Total Charges'].median(), inplace=True)

C:\Users\NIHAR\AppData\Local\Temp\ipykernel_2604\2933346659.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Total Charges'].fillna(df['Total Charges'].median(), inplace=True)


### 📊 Missing Value Handling

- Missing values were found in **Total Charges**  
- These were replaced using the **median value** to avoid skewing the distribution  

This ensures the dataset remains consistent for modeling.

In [86]:
# Drop potential data leakage columns
df = df.drop(columns=['CLTV'])

### ⚠️ Data Leakage Prevention

- **CLTV (Customer Lifetime Value)** may include future information  
- Keeping it can artificially inflate model performance  

Therefore, it is removed to ensure realistic predictions.

In [87]:
df.shape

(7043, 24)

In [88]:
df.head()

,City,Zip Code,Latitude,Longitude,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn
0,Los Angeles,90003,33.964131,-118.272783,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,1
1,Los Angeles,90005,34.059281,-118.307420,Female,No,No,Yes,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,1
2,Los Angeles,90006,34.048013,-118.293953,Female,No,No,Yes,8,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,1
3,Los Angeles,90010,34.062125,-118.315709,Female,No,Yes,Yes,28,Yes,Yes,Fiber optic,No,No,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,1
4,Los Angeles,90015,34.039224,-118.266293,Male,No,No,Yes,49,Yes,Yes,Fiber optic,No,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,1


### ✅ Final Dataset Status

- Dataset is now cleaned and ready for analysis  
- Irrelevant and redundant features removed  
- Target variable properly defined  
- Missing values handled  

Next step: Exploratory Data Analysis (EDA)

---

## 📊 Step 5: Exploratory Data Analysis (EDA)

In this section, we analyze customer behavior to identify patterns associated with churn.

The goal is to:
- Understand key drivers of churn  
- Identify high-risk customer segments  
- Extract actionable business insights  

In [89]:
df['Churn'].value_counts(normalize=True)

Churn
0    0.73463
1    0.26537
Name: proportion, dtype: float64

In [90]:
numeric_cols = ['Tenure Months', 'Monthly Charges', 'Total Charges', 'CLTV']

for col in numeric_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outlier_rows = df[(df[col] < lower) | (df[col] > upper)]

    print(f"{col}: {len(outlier_rows)} outliers")

Tenure Months: 0 outliers
Monthly Charges: 0 outliers
Total Charges: 0 outliers


KeyError: 'CLTV'

### Outliers
- There are no values below   lower = Q1 - 1.5 * IQR  and there are no value above  upper = Q3 + 1.5 * IQR


In [ ]:
import seaborn as sns

sns.countplot(x='Churn', data=df)
plt.title("Churn Distribution")
plt.show()

df['Churn'].value_counts(normalize=True)

### 📊 Churn Distribution

- The dataset is imbalanced  
- Majority of customers do not churn  
- Minority churn class needs special attention during modeling  

⚠️ This justifies the use of metrics like **F1-score, ROC-AUC instead of accuracy**

In [ ]:
df.groupby('Tenure Months')['Churn'].mean()

In [ ]:
sns.boxplot(x='Churn', y='Tenure Months', data=df)
plt.title("Tenure vs Churn")
plt.show()

### 📊 Tenure vs Churn

- Customers with low tenure have significantly higher churn rates  
- Long-term customers are more stable  

💡 Insight:
- Early-stage customers are high-risk  
- Retention strategies should focus on **first few months**

In [ ]:
df.groupby('Contract')['Churn'].mean().sort_values()

In [ ]:
sns.countplot(x='Contract', hue='Churn', data=df)
plt.title("Contract Type vs Churn")
plt.xticks(rotation=30)
plt.show()

### 📊 Contract Type vs Churn

- Month-to-month customers show highest churn  
- Long-term contracts (1 year / 2 year) reduce churn  

💡 Insight:
- Lock-in contracts improve retention  
- Offering discounts for long-term plans can reduce churn

In [ ]:
df.groupby('Payment Method')['Churn'].mean().sort_values()

### 📊 Payment Method vs Churn

- Electronic check users have higher churn  
- Automatic payment methods show lower churn  

💡 Insight:
- Customers using manual payment methods are less committed  
- Incentivizing auto-pay could reduce churn

In [ ]:
df.groupby('Churn')['Monthly Charges'].mean()

In [ ]:
sns.boxplot(x='Churn', y='Monthly Charges', data=df)
plt.title("Monthly Charges vs Churn")
plt.show()

### 📊 Monthly Charges vs Churn

- Customers with higher monthly charges tend to churn more  

💡 Insight:
- High pricing may be a churn driver  
- Price optimization or bundled offers may help retention

In [ ]:
df.groupby('Internet Service')['Churn'].mean()

In [ ]:
sns.countplot(x='Internet Service', hue='Churn', data=df)
plt.title("Internet Service vs Churn")
plt.show()

### 📊 Internet Service vs Churn

- Fiber optic users show higher churn compared to DSL  

💡 Insight:
- Premium services may come with higher expectations  
- Poor experience → higher churn

In [ ]:
sns.boxplot(x='Churn', y='num_services', data=df)
plt.title("Number of Services vs Churn")
plt.show()

In [ ]:
services = [
    'Online Security', 'Online Backup', 'Device Protection',
    'Tech Support', 'Streaming TV', 'Streaming Movies'
]

for col in services:
    print(col)
    print(df.groupby(col)['Churn'].mean())
    print()

### 📊 Service Features vs Churn

- Customers without additional services (security, backup, support) churn more  
- Customers using multiple services are more loyal  

💡 Insight:
- Higher service engagement reduces churn  
- Bundling services can improve retention

In [ ]:
df.groupby('Senior Citizen')['Churn'].mean()

### 📊 Demographics vs Churn

- Senior citizens may have slightly higher churn  

💡 Insight:
- Different customer segments require different retention strategies

In [ ]:
plt.figure(figsize=(8,6))
sns.heatmap(df.select_dtypes(include=np.number).corr(), annot=True, cmap='coolwarm')
plt.title("Correlation Heatmap")
plt.show()

## 🚨 Key Insights from EDA

- Customers with **low tenure** are most likely to churn  
- **Month-to-month contracts** have highest churn rates  
- **High monthly charges** increase churn probability  
- **Electronic check users** show higher churn  
- Customers without **additional services** are more likely to leave  

---

## 💡 Business Takeaways

- Focus retention efforts on **new customers**  
- Promote **long-term contracts**  
- Encourage **auto-pay methods**  
- Bundle services to increase engagement  

---

## ⚙️ Step 6: Feature Engineering

In this step, we create new features based on insights from exploratory data analysis (EDA).

The goal is to:
- Improve model performance  
- Capture hidden patterns in customer behavior  
- Translate business insights into measurable variables  

In [ ]:
df['zip_group'] = pd.qcut(df['Zip Code'], q=5)

### 📍 Zip Code Grouping

- Zip codes were grouped into 5 quantile-based buckets  
- This was done to reduce high cardinality  

### ❌ Observation

- Churn rates across zip groups are almost identical  
- No meaningful pattern observed  

### ⚠️ Conclusion

- This feature does not contribute to churn prediction  
- It will be **excluded from modeling**

In [ ]:
services = [
    'Online Security', 'Online Backup', 'Device Protection',
    'Tech Support', 'Streaming TV', 'Streaming Movies'
]

df['num_services'] = (df[services] == 'Yes').sum(axis=1)

### 🧩 Number of Services

- Counts how many additional services a customer is using  

### 💡 Motivation

EDA showed:
- Customers using more services are less likely to churn  

### 🎯 Insight

- Higher engagement → lower churn  
- This feature captures **customer dependency on services**

In [ ]:
df['Tenure_group'] = pd.cut(
    df['Tenure Months'],
    bins=[0, 12, 24, 48, 72],
    labels=['0-1yr', '1-2yr', '2-4yr', '4-6yr']
)

### ⏳ Tenure Grouping

- Customers are grouped based on their tenure  

### 💡 Motivation

EDA showed:
- New customers churn more frequently  

### 🎯 Insight

- Early-stage customers are high risk  
- This feature helps model capture lifecycle stages

In [ ]:
df['AvgCharges'] = df['Total Charges'] / df['Tenure Months']
df['AvgCharges'].replace([np.inf, -np.inf], 0, inplace=True)

### 💰 Average Charges per Month

- Derived as Total Charges divided by tenure  

### 💡 Motivation

- Monthly Charges alone may not reflect long-term spending  
- This captures **true customer value over time**

### 🔥 Importance

- Identified as the **most important feature (SHAP analysis)**  

### 🎯 Insight

- Customers paying more over time may have higher churn risk

In [ ]:
df['HighRisk_Combo'] = (
    (df['Contract'] == 'Month-to-month') &
    (df['Monthly Charges'] > df['Monthly Charges'].median())
).astype(int)

### ⚠️ High Risk Combination

- Identifies customers with:
  - Month-to-month contracts  
  - High monthly charges  

### 💡 Motivation

EDA showed:
- These two factors strongly increase churn  

### 🎯 Insight

- Combined risk is more powerful than individual features  
- Helps model capture **high-risk customer segments**

In [ ]:
df['ServiceIntensity'] = df['num_services'] / len(services)

df['ServiceIntensity'] = df['num_services'] / len(services)

### 📊 Service Intensity

- Ratio of services used by the customer  

### 💡 Motivation

- Normalizes service usage  

### 🎯 Insight

- Higher intensity → more engagement → lower churn

In [ ]:
df = df.drop(columns=['zip_group'])

## 🚨 Feature Engineering Summary

Created features:

- **num_services** → customer engagement  
- **Tenure_group** → lifecycle stage  
- **AvgCharges** → customer value over time  
- **HighRisk_Combo** → high-risk behavior pattern  
- **ServiceIntensity** → normalized service usage  

---

## 💡 Key Takeaway

Feature engineering helped:

- Capture customer behavior patterns  
- Improve model interpretability  
- Translate business insights into measurable variables  

---

## 🚨 Feature Engineering Summary

Created features:

- **num_services** → customer engagement  
- **Tenure_group** → lifecycle stage  
- **AvgCharges** → customer value over time  
- **HighRisk_Combo** → high-risk behavior pattern  
- **ServiceIntensity** → normalized service usage  

---

## 💡 Key Takeaway

Feature engineering helped:

- Capture customer behavior patterns  
- Improve model interpretability  
- Translate business insights into measurable variables  

In [ ]:
# Separate numerical and categorical columns

numerical_cols = [
    'Tenure Months', 'Monthly Charges', 'Total Charges',
    'num_services', 'AvgCharges', 'ServiceIntensity'
]

categorical_cols = [
    'Gender', 'Senior Citizen', 'Partner', 'Dependents',
    'Phone Service', 'Multiple Lines', 'Internet Service',
    'Online Security', 'Online Backup', 'Device Protection',
    'Tech Support', 'Streaming TV', 'Streaming Movies',
    'Contract', 'Paperless Billing', 'Payment Method',
    'HighRisk_Combo'
]

In [ ]:
df.columns

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Focusing on your specific numerical features
cols_to_check = [
    'Tenure Months', 'Monthly Charges', 'Total Charges', 
    'num_services', 'AvgCharges', 'ServiceIntensity'
]

plt.figure(figsize=(10, 8))
correlation_matrix = df[cols_to_check].corr()

sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', fmt=".2f", linewidths=0.5)
plt.title("Correlation Heatmap of Numerical Features")
plt.show()

### 📊 Feature Categorization

- Numerical features include continuous variables such as charges and tenure  
- Categorical features include customer attributes and service usage  

Separating them allows applying appropriate transformations.

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

num_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

### 🔢 Numerical Feature Processing

- Missing values handled using **median imputation**  
- Features scaled using **StandardScaler**

💡 Why scaling?
- Models like Logistic Regression are sensitive to feature scale  
- Helps improve convergence and performance

In [ ]:
from sklearn.preprocessing import OneHotEncoder

cat_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('encoder', OneHotEncoder(drop='if_binary', handle_unknown='ignore'))
])

### 🔤 Categorical Feature Processing

- Missing values filled using most frequent category  
- One-hot encoding used to convert categories into numerical format  

💡 Why OneHotEncoder?
- Avoids ordinal assumptions  
- Prevents unintended relationships between categories  

In [ ]:
from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer([
    ('num', num_pipeline, numerical_cols),
    ('ohe', cat_pipeline, categorical_cols)
])

### 🔄 Column Transformer

- Applies numerical transformations to numerical features  
- Applies categorical transformations to categorical features  

Ensures each feature type is processed correctly.

In [ ]:

print("Preprocessing pipeline ready.")

### 🚀 Why Pipeline Matters

Using a pipeline ensures:

- No data leakage (transformations applied after split)  
- Reproducibility of results  
- Easy integration with model training and tuning  

This is a best practice in production-level machine learning workflows.

- This pipeline can be directly reused in production environments.

## ✂️ Step 8: Train-Test Split

We split the dataset into training and testing sets to evaluate model performance on unseen data.

This helps ensure that the model generalizes well and avoids overfitting.

In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop('Churn', axis=1)
y = df['Churn']

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

### 📊 Why Stratified Split?

- Ensures same churn distribution in train and test sets  
- Important for imbalanced datasets  

Without this, evaluation can be misleading.

## 🤖 Step 9: Baseline Model — Logistic Regression

We start with Logistic Regression as a baseline model because:

- It is simple and interpretable  
- Works well for binary classification  
- Provides insights into feature importance  

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline

logistic_model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(
        max_iter=1000,
        class_weight='balanced',
        penalty='l1',
        solver='liblinear'
    ))
])

### ⚖️ Handling Class Imbalance

- Dataset is imbalanced (fewer churn cases)  
- Used `class_weight='balanced'` to give more importance to minority class  

This helps improve recall for churn prediction.

## ⚙️ Step 9.1: Hyperparameter Tuning (Logistic Regression)

We tune key parameters to improve model performance:

- Regularization strength (C)  
- Penalty (L1 / L2)  

In [ ]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'classifier__C': [0.01, 0.1, 1, 10],
    'classifier__penalty': ['l1', 'l2']
}

grid = GridSearchCV(
    logistic_model,
    param_grid,
    cv=5,
    scoring='roc_auc',
    n_jobs=-1
)

grid.fit(X_train, y_train)

print("Best Params:", grid.best_params_)
print("Best Score:", grid.best_score_)

In [ ]:
logistic_model = grid.best_estimator_

In [ ]:
# Fit the logistic pipeline on the training set
logistic_model.fit(X_train, y_train)

print("Logistic model trained.")

## 🔁 Step 11: Cross-Validation

We evaluate model stability using 5-fold cross-validation.

Metrics:
- ROC-AUC
- PR-AUC
- F1-score

In [ ]:
from sklearn.model_selection import cross_validate

scoring = {
    'roc_auc': 'roc_auc',
    'pr_auc': 'average_precision',
    'f1': 'f1'
}

cv_results = cross_validate(
    logistic_model,
    X_train,
    y_train,
    cv=5,
    scoring=scoring,
    n_jobs=-1
)

print("Mean ROC-AUC:", cv_results['test_roc_auc'].mean())
print("Mean PR-AUC:", cv_results['test_pr_auc'].mean())
print("Mean F1:", cv_results['test_f1'].mean())

## 📊 Step 12: Evaluation on Test Set

We evaluate performance on unseen data.

In [ ]:
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    roc_auc_score,
    average_precision_score
)

y_pred = logistic_model.predict(X_test)
y_prob = logistic_model.predict_proba(X_test)[:, 1]

print("ROC-AUC:", roc_auc_score(y_test, y_prob))
print("PR-AUC:", average_precision_score(y_test, y_prob))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

## 🎯 Step 13: Threshold Optimization

We adjust probability threshold to improve recall for churn prediction.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.metrics import precision_score, recall_score, f1_score

thresholds = np.arange(0.2, 0.8, 0.05)

results = []

for t in thresholds:
    y_pred_t = (y_prob >= t).astype(int)
    
    results.append([
        t,
        precision_score(y_test, y_pred_t),
        recall_score(y_test, y_pred_t),
        f1_score(y_test, y_pred_t)
    ])

threshold_df = pd.DataFrame(
    results,
    columns=['Threshold', 'Precision', 'Recall', 'F1']
)

threshold_df

## 🤖 Step 14: Model Comparison

A strong churn project should not stop at one model.

To evaluate whether Logistic Regression is sufficient, we compare it with two additional models:

- **Logistic Regression** → interpretable baseline
- **Random Forest** → captures non-linear relationships
- **XGBoost** → high-performance boosting model widely used in industry

This comparison helps determine whether we should prioritize:

- **Interpretability** for business explainability
- **Performance** for stronger churn detection

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.pipeline import Pipeline
from sklearn.metrics import roc_auc_score, average_precision_score, f1_score

models = {
    "Logistic Regression": logistic_model,
    "Random Forest": Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', RandomForestClassifier(
            n_estimators=300,
            max_depth=None,
            min_samples_split=5,
            min_samples_leaf=2,
            random_state=42,
            n_jobs=-1
        ))
    ]),
    "XGBoost": Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', XGBClassifier(
            n_estimators=300,
            learning_rate=0.05,
            max_depth=4,
            subsample=0.8,
            colsample_bytree=0.8,
            eval_metric='logloss',
            random_state=42
        ))
    ])
}

comparison_results = []

for name, model in models.items():
    model.fit(X_train, y_train)
    y_prob_model = model.predict_proba(X_test)[:, 1]
    y_pred_model = model.predict(X_test)

    comparison_results.append({
        "Model": name,
        "ROC-AUC": roc_auc_score(y_test, y_prob_model),
        "PR-AUC": average_precision_score(y_test, y_prob_model),
        "F1": f1_score(y_test, y_pred_model)
    })

comparison_df = pd.DataFrame(comparison_results).sort_values(by="ROC-AUC", ascending=False)
comparison_df

### 🔍 Model Comparison Results

| Model               | ROC-AUC | PR-AUC | F1   |
|--------------------|--------|--------|------|
| XGBoost            | 0.855  | 0.667  | 0.590 |
| Logistic Regression| 0.847  | 0.638  | 0.619 |
| Random Forest      | 0.844  | 0.655  | 0.574 |

---

### ⚖️ Key Takeaways

- **XGBoost** achieves the highest ROC-AUC and PR-AUC → best ranking ability  
- **Logistic Regression** has the highest F1-score → better balance  
- **Random Forest** underperforms compared to others  

---

### 🧠 Final Decision

Logistic Regression is selected as the final model because:

- Performance is **very close to XGBoost**
- More **interpretable**
- Easier to explain to business stakeholders  

---

### 💼 Business Perspective

In churn prediction:

- Slight performance gains are often **not worth losing interpretability**
- Stakeholders prefer models they can understand and trust  

👉 Logistic Regression provides **actionable insights + strong performance**

## 🏆 Step 15: Final Model Selection

After comparing multiple models, we now choose the final model.

The selection is based on:
- Predictive performance
- Ability to detect churners effectively
- Business interpretability
- Practical usability in a retention workflow

In [ ]:
# Replace this manually after checking comparison_df
final_model = logistic_model
final_model_name = "Logistic Regression"

print("Selected Final Model:", final_model_name)

### ✅ Final Model Decision

The final model should be selected based on the comparison results above.

#### Use Logistic Regression if:
- performance is close to the best model
- explainability matters more
- business stakeholders need transparent reasoning

#### Use XGBoost if:
- it clearly outperforms other models
- maximizing churn detection is the top priority
- interpretability can be handled through SHAP explanations

## 🧠 Step 16: Model Explainability (SHAP)

Prediction alone is not enough.

To make the model useful in a real business setting, we need to understand **why** customers are predicted to churn.

SHAP helps explain:
- which features influence churn the most
- whether the model aligns with business intuition
- whether the results are trustworthy enough for action

In [ ]:
import shap

# Fit final model if not already fitted
final_model.fit(X_train, y_train)

# Transform training data
X_train_transformed = final_model.named_steps['preprocessor'].transform(X_train)

# Feature names after preprocessing
feature_names = final_model.named_steps['preprocessor'].get_feature_names_out()

# Extract classifier
classifier = final_model.named_steps['classifier']

# Build SHAP explainer
explainer = shap.Explainer(classifier, X_train_transformed, feature_names=feature_names)
shap_values = explainer(X_train_transformed)

# Beeswarm plot
shap.plots.beeswarm(shap_values, max_display=15)

### 🔍 SHAP Analysis Interpretation

The SHAP summary plot highlights the most important drivers of churn:

#### 🔥 Top Drivers Increasing Churn Risk
- High **Monthly Charges**
- **Month-to-month contracts**
- **Fiber optic internet**
- Lower **Tenure**

#### 🧊 Factors Reducing Churn
- Longer tenure customers  
- Two-year contracts  
- Customers with dependents  

---

### 🧠 Insight

- Pricing and contract type are the strongest churn drivers  
- Customer loyalty (tenure) significantly reduces churn  

---

### 💼 Business Meaning

This confirms:

- Customers paying more expect better value  
- Lack of commitment (monthly contracts) increases churn risk  

👉 The model aligns with real-world business intuition → **high trust model**

## 📊 Step 17: Threshold Optimization

The default classification threshold is **0.50**, but that may not be ideal for churn prediction.

In churn use cases:
- Missing a churner is expensive
- It is often better to increase recall
- Businesses may accept more false positives if it helps save more real churners

So instead of blindly using 0.50, we evaluate multiple thresholds.

In [ ]:
import numpy as np
from sklearn.metrics import precision_score, recall_score, f1_score

thresholds = np.arange(0.20, 0.81, 0.05)
threshold_results = []

y_prob_final = final_model.predict_proba(X_test)[:, 1]

for t in thresholds:
    y_pred_t = (y_prob_final >= t).astype(int)
    threshold_results.append({
        "Threshold": round(t, 2),
        "Precision": precision_score(y_test, y_pred_t),
        "Recall": recall_score(y_test, y_pred_t),
        "F1": f1_score(y_test, y_pred_t),
        "Flagged_Customers": y_pred_t.sum()
    })

threshold_df = pd.DataFrame(threshold_results)
threshold_df

In [ ]:
best_f1_row = threshold_df.loc[threshold_df["F1"].idxmax()]
best_f1_row

### 🎯 Optimal Threshold Selection

| Threshold | Precision | Recall | F1 |
|----------|----------|--------|----|
| 0.50     | 0.509    | 0.791  | 0.619 |
| **0.60** | **0.562**| **0.719** | **0.631** |

---

### ✅ Final Threshold: **0.60**

---

### 🔍 Why 0.60?

- Improves **precision** (fewer unnecessary interventions)  
- Maintains strong **recall (~72%)**  
- Achieves best **F1-score**

---

### 💼 Business Decision Logic

- At 0.50 → too many customers flagged  
- At 0.60 → better balance between cost and effectiveness  

👉 This ensures:
- Efficient use of retention budget  
- High-quality targeting  .

## 💰 Step 18: Business Impact Metrics

A strong churn project should not stop at model metrics.

Business teams care about questions like:

- How many customers will be flagged?
- How many real churners can we potentially save?
- Is the retention campaign operationally feasible?

This section translates model output into business language.

In [ ]:
# Use threshold = 0.50 by default, or replace with chosen threshold
chosen_threshold = 0.50
y_pred_business = (y_prob_final >= chosen_threshold).astype(int)

tn, fp, fn, tp = confusion_matrix(y_test, y_pred_business).ravel()

total_test_customers = len(y_test)
flagged_customers = y_pred_business.sum()
actual_churners = y_test.sum()
captured_churners = tp

print("Chosen Threshold:", chosen_threshold)
print("Total customers in test set:", total_test_customers)
print("Customers flagged as high risk:", flagged_customers)
print("Actual churners in test set:", actual_churners)
print("Correctly identified churners (TP):", captured_churners)
print("False positives:", fp)
print("Recall-based churn capture:", captured_churners / actual_churners)
print("Targeting rate:", flagged_customers / total_test_customers)

In [ ]:
business_summary = pd.DataFrame({
    "Metric": [
        "Total test customers",
        "Flagged as high risk",
        "Actual churners",
        "Captured churners",
        "False positives",
        "Churn capture rate",
        "Targeting rate"
    ],
    "Value": [
        total_test_customers,
        flagged_customers,
        int(actual_churners),
        int(captured_churners),
        int(fp),
        round(captured_churners / actual_churners, 3),
        round(flagged_customers / total_test_customers, 3)
    ]
})

business_summary

### 💼 Business Impact Analysis

- Total customers evaluated: **1409**
- Customers flagged as high risk: **582 (~41%)**
- Actual churners: **374**
- Correctly identified churners: **296**

---

### 📊 Key Metrics

- **Churn Capture Rate:** 79%  
- **Targeting Rate:** 41%  

---

### 🔍 Interpretation

- Model successfully captures **~80% of churners**
- Requires targeting **~41% of customers**

---

### 💡 Business Trade-off

- High recall ensures fewer churners are missed  
- Moderate false positives are acceptable in retention campaigns  

👉 This is **practically deployable**


## 🎯 Step 19: High-Risk Customer Segmentation

Instead of treating all customers equally, we can rank customers by churn probability.

This helps the business:
- prioritize outreach
- focus on the highest-risk customers first
- allocate retention budget more effectively

In [ ]:
risk_df = X_test.copy()
risk_df["Actual_Churn"] = y_test.values
risk_df["Churn_Probability"] = y_prob_final

risk_df = risk_df.sort_values(by="Churn_Probability", ascending=False)

risk_df[["Churn_Probability", "Actual_Churn"]].head(20)

In [ ]:
top_20_percent = int(len(risk_df) * 0.20)
top_risk_segment = risk_df.head(top_20_percent)

captured_in_top20 = top_risk_segment["Actual_Churn"].sum()
total_churners_test = risk_df["Actual_Churn"].sum()

print("Top 20% customers count:", top_20_percent)
print("Churners captured in top 20%:", int(captured_in_top20))
print("Capture rate in top 20%:", captured_in_top20 / total_churners_test)

### 🎯 Top 20% Customer Targeting Strategy

- Top 20% customers analyzed: **281**
- Churners captured: **188**

---

### 📊 Result

- **~50% of all churners captured by targeting just 20% customers**

---

### 💼 Business Value

This is extremely powerful:

- Instead of targeting 41% customers  
- Business can focus on **top 20% high-risk segment**

---

### 🚀 Impact

- Reduces cost of retention campaigns  
- Improves efficiency  
- Enables prioritized outreach  

👉 This is a **production-level insight**

### 📈 Final Business Insights

#### 🚨 High-Risk Customers
- Month-to-month contracts  
- High monthly charges  
- Fiber optic users  
- Low tenure customers  

#### 🧊 Low-Risk Customers
- Long-term contracts  
- Long tenure  
- Customers with dependents  

---

### 💡 Recommended Actions

- Incentivize long-term contracts  
- Improve service quality for high-paying users  
- Target early-stage customers with retention offers  

### 📈 Final Business Insights

#### 🚨 High-Risk Customers
- Month-to-month contracts  
- High monthly charges  
- Fiber optic users  
- Low tenure customers  

#### 🧊 Low-Risk Customers
- Long-term contracts  
- Long tenure  
- Customers with dependents  

---

### 💡 Recommended Actions

- Incentivize long-term contracts  
- Improve service quality for high-paying users  
- Target early-stage customers with retention offers  